jax/experimental/sparse/bcoo.py

```diff
@@ -610,6 +614,136 @@ def _bcoo_dot_general_abstract_eval(lhs_data, lhs_indices, rhs, *, dimension_num
 
   return core.ShapedArray(out_shape, lhs_data.dtype)
 
+def _bcoo_padding_indices_correction_impl(bcoo_indices, data_shape):
+  """Replaces out-of-bound indices in a padded representation with zeros.
+
+  Assume the input has no batch nor dense dimension and has already passed
+  `validate_bcoo` and `validate_bcoo_indcies`.
+
+  Here is an example of padded representation:
+  bcoo_mat = sparse.BCOO.fromdense(jnp.array([[1, 0, 2, 3, 0],
+                                              [0, 0, 0, 4, 0]]), nse=7)
+  has `bcoo_data = [1 2 3 4 0 0 0]` and
+  `bcoo_indices = [[0 0], [0 2], [0 3], [1 3], [2 5], [2 5], [2 5]]`. The actual
+  number of nonzeros is `4` whereas the user-defined `nse` is `7` such that
+  padding occurs.
+
+  The correction creates new `bcoo_indices` such that
+  `bcoo_indices = [[0 0], [0 2], [0 3], [1 3], [0 0], [0 0], [0 0]]`, where
+  all `[2 5]`s are replaced with `[0 0]`.
+
+  The returned array contain indices of type `jnp.int32`.
+  """
+  # TODO: support batch dimension?
+  assert len(data_shape) <= 2
+  assert len(bcoo_indices.shape) <= 2
+
+  nse, n_sparse = bcoo_indices.shape
+  assert n_sparse <= 2
+
+  mask = (jnp.broadcast_to(data_shape, (nse, n_sparse)) -
+          bcoo_indices).astype(bool).astype(jnp.int32)
+
+  bcoo_indices = jnp.multiply(bcoo_indices, mask)
```

Comment:
In order to catch *any* out-of-bound index, we could do something like this:
```python
bcoo_indices = jnp.where(bcoo_indices < jnp.array(data_shape), bcoo_indices, 0)
```

Reply:
Changed to `jnp.where`.  I think `jnp.where` is based on `lax.select`, `broadcast_arrays`, and complex logic like element-wise not equal `lax.ne`, which may not be as efficient as the matrix multiplications on GPU/TPU.

Reply:
Sure, but the matrix multiplication approach only catches out-of-bound indices that are exactly equal to the shape, and will lead to a segfault otherwise. I think we should focus on avoiding segfaults before focusing on optimization!

Reply:
Agree!